### DISCUSSION TITLE
How to test using vitest

### DISCUSSION THREAD

**Discussion by charklewis** [fastify/fastify#3838] [Q&A]
I would like to test my fastify app using [vitest](https://vitest.dev), however I have noticed there is a typescript error I can't seem to get around.

When I run my tests Fastify errors, `fastify-autoload cannot import plugin ... . To fix this error compile TypeScript to JavaScript or use 'ts-node' to run your app.`. 

Does this mean I have to run my tests based on the javascript versions of my tests? I am sure there would be a way to configure fastify to work with vitest using the typescript files. 

My vitest configuration looks like the below:

```
/// <reference types="vitest" />
/// <reference types="vite/client" />

import { defineConfig } from "vite"
import tsconfigPaths from "vite-tsconfig-paths"

export default defineConfig({
  plugins: [tsconfigPaths()]
})
```

**Comment by mcollina** [comment]
Maybe @galvez could help?

**Comment by charklewis** [comment]
I have a feeling that I would need to get `vitest` to compile the code to javascript and use that for its tests.

**Reply by mcollina** [reply]
I don't know enough about the internals of vitest to know. However I think you'll need to send a patch to fastify-autoload.

**Reply by galvez** [reply]
Hi @charklewis I'm currently looking into potential Vitest integration, but it's hard to make a case for it when we don't have to bundle Fastify server-side code through Vite.

**Reply by mcollina** [reply]
@galvez I think the actual problem is simpler: fastify-autoload must be aware of how the transpiled code is loaded. We do the following for jest and others: https://github.com/fastify/fastify-autoload/blob/56fa01a601ec0ca4fc6db3313b708bdbe95462b3/index.js#L10-L12.

**Reply by charklewis** [reply]
Could a patch be as simple as checking for if Vitest is running the code and adding it to the `typescriptSupport` check?

**Reply by mcollina** [reply]
Possibly, I don't know how vitest work internally.

**Reply by charklewis** [reply]
As a test I ran `JEST_WORKER_ID="1" yarn test` (to force `typescriptSupport=true`) and it looks like it is erroring now as I am using ES6 import in my file `import { FastifyPluginCallback } from "fastify"`.

**Reply by charklewis** [reply]
I can test my routes by importing the compiled JS code (instead of the typescript code) but it does feel a bit hacky. So this isn't an immediate blocker 🙂

I feel it will likely be a problem in the near future as vitest [continues to increase in popularity](https://www.npmtrends.com/vitest).

**Reply by mcollina** [reply]
If somebody is interested, they'll send a contribution.

**Reply by mcollina** [reply]
Could you open an issue on fastify-autoload? So we can track it as a good first issue?

**Reply by charklewis** [reply]
I can create an issue tonight with a sandbox example 🙂 I don't mind having a go at patching this but wouldn't know where to start.

**Reply by charklewis** [reply]
@mcollina @galvez I have created an issue on the repo with a stackblitz example - https://github.com/fastify/fastify-autoload/issues/230.

**Comment by rdrigos** [comment]
Good morning, I am having trouble testing my application with Vitest and Fastify. I have this class that I created, the server:

```
import { router } from '@/http/router';
import Fastify, { FastifyInstance } from 'fastify';

interface ServerOptions {
  host: string;
  port: number;
}

export class Server {
  private options: ServerOptions;
  private server: FastifyInstance;

  constructor(options: ServerOptions) {
    this.options = options;
    this.server = Fastify({ logger: false });

    // Configure the server
    this.configureServer();
  }

  private configureServer(): void {
    this.server.register(router, { prefix: '/' });
  }

  public async start(): Promise<void> {
    return this.server
      .listen({
        host: this.options.host,
        port: this.options.port,
      })
      .then((address: string) => {
        console.log(`HTTP server started successfully at: ${address}`);
      })
      .catch((error: any) => {
        console.error(`HTTP server encountered an error during startup: ${error}`);
      });
  }
}
```

And these are the tests (for now, I only created the success test...):

```
import { router } from '@/http/router';
import { Server } from '@/http/server';
import { faker } from '@faker-js/faker';
import Fastify, { FastifyInstance } from 'fastify';
import { beforeEach, describe, expect, it, Mock, vi } from 'vitest';

vi.mock('fastify');
vi.mock('@/http/router');

describe('Server', () => {
  let serverInstance: FastifyInstance;
  let consoleSpy;
  let options: any;

  beforeEach(() => {
    // Mock of the FastifyInstance
    serverInstance = {
      register: vi.fn(),
      listen: vi.fn(() => Promise.resolve()),
    } as unknown as FastifyInstance;

    // Ensure that Fastify always returns the mocked instance
    (Fastify as unknown as Mock).mockReturnValue(serverInstance);

    // Spy on console.log
    consoleSpy = vi.spyOn(global.console, 'log');
  });

  it('It should configure and start the server correctly', async () => {
    options = {
      host: faker.internet.domainName(),
      port: faker.number.int({ min: 80, max: 65535 }),
    };

    // Create a server instance
    const server = new Server(options);

    // Check if the router was registered correctly
    expect(serverInstance.register).toHaveBeenCalledWith(router, { prefix: '/' });

    // Start the server
    server.start();

    // Check if the server started on the correct port
    expect(serverInstance.listen).toHaveBeenCalledWith({ host: options.host, port: options.port });

    // Check if console.log was called
    expect(consoleSpy).toHaveBeenCalledWith(expect.stringContaining('HTTP server started successfully at:'));
  });
});
```

It is not able to validate that console.log was called. Why?
